### ISSUE TITLE
Feature Request: Zerotier

### ISSUE THREAD

**Issue by mangotek** [sipeed/NanoKVM#79]
Simliar to Tailscale, alternative VPN is Zerotier a SD-WAN

**Comment by Zepan** [comment]
Hi, what is the advantage of Zerotier, compares to Tailscale.
We will add support if it have enough advantages, otherwise you can port by yourself as it is opensource now.

**Comment by amyc935** [comment]
> Hi, what is the advantage of Zerotier, compares to Tailscale. We will add support if it have enough advantages, otherwise you can port by yourself as it is opensource now.

The performance of SG2002 itself is not very powerful.
current Tailscale will take up too much performance, and zerotier will provide a better experience.

**Comment by Pouletgrill** [comment]
I would personally only buy the NanoKVM if it can be integrated in my zero tier environement. I am sure im not the only one out there that use zero tier and dont want to use Tailscale in top of that. 
Thank you.

**Comment by Zepan** [comment]
I see, we will consider it

**Comment by shimco** [comment]
On my remote devices including pikvm's, which are often deployed hundreds of miles away, I use both Tailscale and Zerotier to provide redundancy if one service fails.
Adding multiple options like this would be beneficial for the NanoKVM project, so I hope you can implement this.

**Comment by staberas** [comment]
---

## **Incomplete and **NOT** yet working Guide to Installing ZeroTier 1.8.1 on NanoKVM **


### **Steps**

#### 1. **Download ZeroTier 1.8.1 `.deb` File**
Use `wget` to fetch the ZeroTier 1.8.1 `.deb` file for `riscv64`:
```bash
mkdir zero
cd zero
wget https://download.zerotier.com/debian/bullseye/pool/main/z/zerotier-one/zerotier-one_1.8.1_riscv64.deb
```

#### 2. **Extract the `.deb` File**
Extract the `.deb` contents using `ar` and `xzcat`:
```bash
ar x zerotier-one_1.8.1_riscv64.deb
xzcat data.tar.xz | tar -xvf -
```

#### 3. **Move Files to Correct Locations**
Move the extracted binaries and files to their appropriate locations:
```bash
cp -r ./usr/sbin/zerotier-one /usr/sbin/
cp -r ./var/lib/zerotier-one /var/lib/
cp -r ./etc/init.d/zerotier-one /etc/init.d/
```


#### 4. **Verify Required Libraries (not sure if 1.8.1 requires them)**
Ensure required libraries like `libstdc++`, `libgcc_s`, `libc6`, and `libm` exist in `/lib` or `/usr/lib`. If missing:
1. Download and extract dependencies from [[Debian repositories](https://ftp.debian.org/debian/pool/main)](https://ftp.debian.org/debian/pool/main).
2. Example for `libc6`:
   ```bash
   wget http://ftp.debian.org/debian/pool/main/g/glibc/libc6_2.40-5_riscv64.deb
   ar x libc6_2.40-5_riscv64.deb
   xzcat data.tar.xz | tar -xvf -
   cp -r ./lib/* /lib/
   ```

ZeroTier requires has an extra dependency in order to fully run which is lsb, up until now i havent managed to fix that and im currently editing zerotier code to sidestep this issue , if anyone with more knowledge want to take on the challenge you can try

**Comment by ciroiriarte** [comment]
would like to see this included, I extensively use Zerotier already.

**Comment by AndrewMoryakov** [comment]
I also actively use Zirotier, it can be much more convenient.

**Comment by GulinSS** [comment]
What to do if it stopped to load after reboot? I mean NanoKVM itself does not loading after these steps, seems I got it broken somehow. 😅

**Comment by staberas** [comment]
> What to do if it stopped to boot after reboot? I mean NanoKVM itself does not loading after these steps, seems I got it broken somehow. 😅

it doesn't boot at all? or zerotier doesn't start at boot ?

**Comment by GulinSS** [comment]
> it doesn't boot at all?

That one. KVM does switch on its display with IP address.

**Comment by staberas** [comment]
you have to reflash the sd card, but before that you need to see if the device is available on the network and access it.

**Comment by GulinSS** [comment]
Yes, reflasing internal sd card with fresh image fixed the device. Well, another turn to enable ZeroTier :-)

**Comment by staberas** [comment]
> [@staberas](https://github.com/staberas) which versions did you use?

![Image](https://github.com/user-attachments/assets/10dde90b-b779-4164-8d94-e5bee22aa349)

1.8.1 , its the only one i got working . (i started with 1.14.something added the libs on step 4 but it failed to work then went down to 1.8.1 which i know works on the licheeRV )

**Comment by GulinSS** [comment]
`1.14` worked for me but I got a broken state after kvm reboot: it was unable to connect to network and turn on lcd.

**Comment by staberas** [comment]
@GulinSS  
So apparently 
 
has been replaced by 
http://ftp.debian.org/debian/pool/main/g/glibc/libc6_2.40-6_riscv64.deb
(i have the old one)

regardless

post me what /usr/sbin/zerotier-one tells you


and do the following :
`strings /usr/sbin/zerotier-one | grep '\.so'  `

you should get the following :

```
/lib/ld-linux-riscv64-lp64d.so.1
libstdc++.so.6
libm.so.6
libgcc_s.so.1
libpthread.so.0
libc.so.6
ld-linux-riscv64-lp64d.so.1
/var/run/minissdpd.sock
```

After downloading it should do the following (confirming):

` ar x libc6_2.40-5_riscv64.deb`  or  ` ar x libc6_2.40-6_riscv64.deb `
 
//we dont need debian-binary
` rm control.tar.xz debian-binary 
`
//decompress and move the files 
```
 xzcat data.tar.xz | tar -xvf - 
 mv usr/lib/* /lib/ 

```
//remove the data.tar when done
`  rm -rf usr lib data.tar.xz `

**Comment by Josephur** [comment]
Thanks a lot guys! I was able to get zerotier working on my nanovkm using latest firmware, and by using the directions above but using the latest libc from http://ftp.debian.org/debian/pool/main/g/glibc/libc6_2.41-1_riscv64.deb

However I'm not sure if I missed something in the instructions but for me the usr/lib/zerotier-cli and usr/binzerotier-idtool binaries that came in the deb file did not copy, I just copied them manually to /usr/sbin

This really needs to be a built in feature to choose between zerotier and/or tailscale.

**Comment by Josephur** [comment]
@staberas I'm going to flash back to stock and make a YT video actually, as I see one does not yet exist.  I will provide the commands I use here as well.  I hope to get it done in the next few days.

**Comment by yesm1ke** [comment]
Just preordered nanokvm and can't wait for delivery and zerotier support out of box!

Great product, thanks guys!

**Comment by Msprg** [comment]
@Josephur 
Hey guys, have you came any closer to making a more complete guide/video about installing zerotier? I tried it once but failed pretty miserably, so I decided to wait since it looked like you were putting some instructions together.

**Comment by GulinSS** [comment]
@Msprg, no, who got it to work and... well.. after that everything more is eaten by laziness.

**Comment by staberas** [comment]
@Msprg which version did you tried to install ?

**Comment by Msprg** [comment]
@staberas  I believe the one mentioned here: "zerotier-one_1.8.1_riscv64.deb"

**Comment by staberas** [comment]
ok, what was the error you encountered, if you can remember...

**Comment by Msprg** [comment]
I can't remember anymore, however, I can try to reproduce it if you'd like

**Comment by staberas** [comment]
please do , just do this [this](https://github.com/sipeed/NanoKVM/issues/79#issuecomment-2599633763) but skip  _4. Verify Required Libraries (not sure if 1.8.1 requires them)_  ( i think those for the 1.14version ) , LSB was needed to make it work.

**Comment by Msprg** [comment]
Okay so...
first of all:
```
wget https://download.zerotier.com/debian/bullseye/pool/main/z/zerotier-one/zerotier-one_1.8.1_riscv64.deb
wget: not an http or ftp url: https://download.zerotier.com/debian/bullseye/pool/main/z/zerotier-one/zerotier-one_1.8.1_riscv64.deb
```

This is fun, curl works...


the rest went technically well, I skipped the step 4 you said to skip


and then I ended up basically here:

```
/zero # /etc/init.d/zerotier-one start
/etc/init.d/zerotier-one: line 25: log_daemon_msg: not found
nice: can't execute '/usr/sbin/zerotier-one': No such file or directory
/etc/init.d/zerotier-one: line 27: log_end_msg: not found
/zero # ls /usr/sbin/zerotier-one 
/usr/sbin/zerotier-one
/zero # chmod +x /usr/sbin/zerotier-one 
/zero # /etc/init.d/zerotier-one start
/etc/init.d/zerotier-one: line 25: log_daemon_msg: not found
nice: can't execute '/usr/sbin/zerotier-one': No such file or directory
/etc/init.d/zerotier-one: line 27: log_end_msg: not found
```


And the last time I ended up here I couldn't make any further progress IIRC...

**Comment by staberas** [comment]
log_daemon_msg , log_end_msg are all LSB dependency https://www.linuxfromscratch.org/lfs/view/development/scripts/apds02.html <- have you added the script ?   
also do an ls on /usr/sbin/ , does  "zerotier-one"  actually exists?

**Comment by Msprg** [comment]
Of course I have added the script, and of course the zerotier binary exists, otherwise the chmod +x command you see above, would fail:

```
/zero # head /lib/lsb/init-functions
#!/bin/sh
########################################################################
#
# Begin /lib/lsb/init-funtions
#
# Description : Run Level Control Functions
#
# Authors     : Gerard Beekmans - gerard@linuxfromscratch.org
#             : DJ Lucas - dj@linuxfromscratch.org
# Update      : Bruce Dubbs - bdubbs@linuxfromscratch.org

/zero # tail /lib/lsb/init-functions
# Purpose: Utility to test if a variable is true | yes | 1                     #
#                                                                              #
################################################################################
is_true()
{
   [ "$1" = "1" ] || [ "$1" = "yes" ] || [ "$1" = "true" ] ||  [ "$1" = "y" ] ||
   [ "$1" = "t" ]
}

# End /lib/lsb/init-functions

/zero # ls /usr/sbin/zerotier*
/usr/sbin/zerotier-one

/zero # file /usr/sbin/zerotier*
/usr/sbin/zerotier-one: ELF 64-bit LSB pie executable, UCB RISC-V, RVC, double-float ABI, version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux-riscv64-lp64d.so.1, BuildID[sha1]=bb0fca64ca755a3382d8bfb8ff1b2c8adbd73b4e, for GNU/Linux 4.15.0, stripped

/zero # 
```

**Comment by staberas** [comment]
ok will check my nanokvm later this evening , i might have done some editing in the zero-tier one to make LSB work

**Comment by Msprg** [comment]
Thank you very much, I would greatly appreciate that~

**Comment by staberas** [comment]
@Msprg 
i had edited the LSB, try replacing with this 
https://gist.github.com/staberas/875fba79e78507f6d87d08a0b852eceb

then try` # zerotier-cli ` or `/usr/sbin/zerotier-cli
`

**Comment by Msprg** [comment]
I did replace the lsb script with yours.

I then tried running zerotier binaries, but all I am getting is this:

```
~ # zero
zero/         zerotier-one
~ # zerotier-one 
/bin/sh: zerotier-one: not found
~ # strace zerotier-one
execve("/sbin/zerotier-one", ["zerotier-one"], 0x3fff979d50 /* 10 vars */) = -1 ENOENT (No such file or directory)
strace: exec: No such file or directory
+++ exited with 1 +++
~ # /usr/sbin/zerotier-one 
/bin/sh: /usr/sbin/zerotier-one: not found
~ # strace /usr/sbin/zerotier-one 
execve("/usr/sbin/zerotier-one", ["/usr/sbin/zerotier-one"], 0x3ffff49d50 /* 10 vars */) = -1 ENOENT (No such file or directory)
strace: exec: No such file or directory
+++ exited with 1 +++
~ # strace ./zero/usr/sbin/zerotier-
zerotier-cli     zerotier-idtool  zerotier-one
~ # strace ./zero/usr/sbin/zerotier-cli 
execve("./zero/usr/sbin/zerotier-cli", ["./zero/usr/sbin/zerotier-cli"], 0x3fffb96d40 /* 10 vars */) = -1 ENOENT (No such file or directory)
strace: exec: No such file or directory
+++ exited with 1 +++
~ # 
```

I am also reminding you that this can very well be explained by missing dynamic libs as you told me to skip step 4 (why?) since it's not really that the system can't find the executables itself which are obviously there, but rather the dynamic library linker cannot find required .so objects.

And on that note, the link http://ftp.debian.org/debian/pool/main/g/glibc/libc6_2.40-6_riscv64.deb is no longer valid, and has been now superseded by http://ftp.debian.org/debian/pool/main/g/glibc/libc6_2.41-9_riscv64.deb

**Comment by staberas** [comment]
> And on that note, the link http://ftp.debian.org/debian/pool/main/g/glibc/libc6_2.40-6_riscv64.deb is no longer valid, and has been now superseded by http://ftp.debian.org/debian/pool/main/g/glibc/libc6_2.41-9_riscv64.deb

yes , it should be that thats missing now 
`# strings /usr/sbin/zerotier-cli | grep '\.so'
`
this will show you what is needed 
```/lib/ld-linux-riscv64-lp64d.so.1
libstdc++.so.6
libm.so.6
libgcc_s.so.1
libpthread.so.0
libc.so.6
ld-linux-riscv64-lp64d.so.1
/var/run/minissdpd.sock
```
here my version in case the new one doesnt work or smthg
[libc6](https://github.com/staberas/libc6_2.40-5_riscv64_/blob/main/libc6_2.40-5_riscv64.deb
)
```
ar x libc6_2.40-5_riscv64.deb
xzcat data.tar.xz | tar -xvf -
cp -r ./lib/* /lib/

```

**Comment by Msprg** [comment]
@staberas Thank you for all the help, seems I've finally got it working then. I've put together an installation script and then put in in a repo so that it's easier to access and maintain it.

Repo here: https://github.com/Msprg/nanoKVM-ZeroTier-one

Or all-on-one release for this thread here: [nanokvm-zerotier.zip](https://github.com/user-attachments/files/20997614/nanokvm-zerotier.zip)

**Comment by Staok** [comment]
https://github.com/sipeed/NanoKVM/issues/79#issuecomment-2599633763

this line `wget http://ftp.debian.org/debian/pool/main/g/glibc/libc6_2.40-5_riscv64.deb` not work.
The newest is: `wget http://ftp.debian.org/debian/pool/main/g/glibc/libc6_2.42-1_riscv64.deb`
From this link: http://ftp.debian.org/debian/pool/main/g/glibc/

Like you should put it like this is better:
`wget http://ftp.debian.org/debian/pool/main/g/glibc/libc6_<the lastest version num>_riscv64.deb`

And!!

the fun "log_daemon_msg" is not in the "init-functions" from  https://www.linuxfromscratch.org/lfs/view/development/scripts/apds02.html

So I use the dummy fun instead.....like below:

```
log_daemon_msg () {
    # Dummy function to be replaced by LSB library.

    echo $@
}
log_progress_msg() {
   echo $@
}
log_end_msg () {
    # Dummy function to be replaced by LSB library.

    if test "$1" != "0"; then
      echo "Error with $DESCRIPTION: $NAME"
    fi
    return $1
}
```

Come on.... linux world is full of this for another half century?
Shall do many cmd things and dependcyxxx things by users? Just write some smart scripts to handle the install and config things.... Is our time not valuable?

**Comment by Ferohers** [comment]
> [#79 (comment)](https://github.com/sipeed/NanoKVM/issues/79#issuecomment-2599633763)
> 
> this line `wget http://ftp.debian.org/debian/pool/main/g/glibc/libc6_2.40-5_riscv64.deb` not work. The newest is: `wget http://ftp.debian.org/debian/pool/main/g/glibc/libc6_2.42-1_riscv64.deb` From this link: http://ftp.debian.org/debian/pool/main/g/glibc/
> 
> Like you should put it like this is better: `wget http://ftp.debian.org/debian/pool/main/g/glibc/libc6_<the lastest version num>_riscv64.deb`
> 
> And!!
> 
> the fun "log_daemon_msg" is not in the "init-functions" from https://www.linuxfromscratch.org/lfs/view/development/scripts/apds02.html
> 
> So I use the dummy fun instead.....like below:
> 
> ```
> log_daemon_msg () {
>     # Dummy function to be replaced by LSB library.
> 
>     echo $@
> }
> log_progress_msg() {
>    echo $@
> }
> log_end_msg () {
>     # Dummy function to be replaced by LSB library.
> 
>     if test "$1" != "0"; then
>       echo "Error with $DESCRIPTION: $NAME"
>     fi
>     return $1
> }
> ```
> 
> Come on.... linux world is full of this for another half century? Shall do many cmd things and dependcyxxx things by users? Just write some smart scripts to handle the install and config things.... Is our time not valuable?

So sad to see how this will break heart of new users...

**Comment by Staok** [comment]
https://github.com/sipeed/NanoKVM/issues/79#issuecomment-3220211203

Oh... I was emotional. Just for when I do these things, Its likely encounter various system environment problems...not just for this, but for the linux things...

**Comment by cvocvo** [comment]
> [@staberas](https://github.com/staberas) Thank you for all the help, seems I've finally got it working then. I've put together an installation script and then put in in a repo so that it's easier to access and maintain it.
> 
> Repo here: https://github.com/Msprg/nanoKVM-ZeroTier-one
> 
> Or all-on-one release for this thread here: [nanokvm-zerotier.zip](https://github.com/user-attachments/files/20997614/nanokvm-zerotier.zip)

Perhaps a dumb question, but in your experience do we need to re-run this after applying application updates on NanoKVM? (I have the full cube version if that makes any material difference.) Thank you

**Comment by staberas** [comment]
@cvocvo 
as far as i have , no , the only thing that might get broken is the boot order number [S97] and at that point you just need to set a new number

**Comment by Msprg** [comment]
> > [@staberas](https://github.com/staberas) Thank you for all the help, seems I've finally got it working then. I've put together an installation script and then put in in a repo so that it's easier to access and maintain it.
> > Repo here: https://github.com/Msprg/nanoKVM-ZeroTier-one
> > Or all-on-one release for this thread here: [nanokvm-zerotier.zip](https://github.com/user-attachments/files/20997614/nanokvm-zerotier.zip)
> 
> Perhaps a dumb question, but in your experience do we need to re-run this after applying application updates on NanoKVM? (I have the full cube version if that makes any material difference.) Thank you

No, I don't believe it will as most if not all updates you're doing from the web ui update just the "main" nano kvm executable. Meaning that, at least for now, it's not really doing extensive system changes.


That said, you will, of course, have to reinstall if you format the sd card and put a new image on it from the releases, for example. Or if they ever issue an update that will do something like that.

**Comment by manueldev** [comment]
> [@staberas](https://github.com/staberas) Thank you for all the help, seems I've finally got it working then. I've put together an installation script and then put in in a repo so that it's easier to access and maintain it.
> 
> Repo here: https://github.com/Msprg/nanoKVM-ZeroTier-one
> 
> Or all-on-one release for this thread here: [nanokvm-zerotier.zip](https://github.com/user-attachments/files/20997614/nanokvm-zerotier.zip)

Installed and working without problems on version 2.2.9.

**Comment by Msprg** [comment]
I'm very happy to hear that! 🥳

**Comment by 1412kauti** [comment]
> [@staberas](https://github.com/staberas) Thank you for all the help, seems I've finally got it working then. I've put together an installation script and then put in in a repo so that it's easier to access and maintain it.
> 
> Repo here: https://github.com/Msprg/nanoKVM-ZeroTier-one
> 
> Or all-on-one release for this thread here: [nanokvm-zerotier.zip](https://github.com/user-attachments/files/20997614/nanokvm-zerotier.zip)

Can Confirm this works with version 2.3.0 as well

**Comment by manueldev** [comment]
I upgraded from 2.2.9 to 2.3.0 and ZeroTier is still working.

**Comment by bulfaitelo** [comment]
I forked the repository and made some changes to make future updates easier, even though I'm currently having problems with ZeroTier versions above 1.8.4.

Here's the repository: https://github.com/bulfaitelo/nanoKVM-ZeroTier-one

Working with nanoKVM version **2.3.6**
